### ISSUE TITLE
Resize varlibqubes pool

### ISSUE THREAD

**Issue by mockfox** [QubesOS/qubes-issues#7184]
[How to file a helpful issue](https://www.qubes-os.org/doc/issue-tracking/)

### The problem you're addressing (if any)
Currently unable to increase the size of the varlibqubes pool. This is an issue as I cannot install more templates since I'm limited by the size of varlibqubes pool.

Also the documentation for storage pools is out of date (https://www.qubes-os.org/doc/storage-pools/). Using R4.1, there are no `storage.conf` files anywhere in dom0.

### The solution you'd like
 There should be an easy configuration option somewhere like configuring the storage size of individual VMs.

### The value to a user, and who that user might be
 This would be useful for anyone who uses a lot of space for templates who runs out of space and cannot install more templates.

**Comment by brendanhoar** [comment]
Under default R4.1 install, the "varlibqubes" pool is listed as using the file driver in qvm-pool, while vms are in the vm-pool, which is lvm_thin driver.

What is varlibqubes pool used for under R4.1?

B

**Comment by DemiMarie** [comment]
@brendanhoar nothing, at least with default partitioning.

**Comment by rustybird** [comment]
Terminology is bound to be confusing here:

- The *storage pool* named varlibqubes (and located in /var/lib/qubes) isn't normally used for anything in the default LVM installation layout.
- But the /var/lib/qubes *directory* is used for other purposes besides hosting a storage pool. Specifically, installing a template will create some large temporary files there.

So @mockfox doesn't really need to resize the varlibqubes pool per se but rather the filesystem at /var/lib/qubes.

**Comment by Kannen** [comment]
I have just installed QubeOS, created 3 qubes, just fall on this at the fourth. Is there really no way to fix this issue as a user?

**Comment by brendanhoar** [comment]
> So @mockfox doesn't really need to resize the varlibqubes pool per se but rather the dom0 filesystem at /var/lib/qubes.

Assuming default install of LVM, this (growing the dom0 pool and filesystem) should be happening automatically on reboot after the autoextend settings/procedure that Marek added to R4.1, yes?

Though, on the LVM side, the values used *might not be big enough* to make room for a large, full-featured template?
From R4.1's ```/etc/lvm/lvm.conf```:

```
	# Configuration option activation/thin_pool_autoextend_threshold.
	# Auto-extend a thin pool when its usage exceeds this percent.
	# Setting this to 100 disables automatic extension.
	# The minimum value is 50 (a smaller value is treated as 50.)
	# Also see thin_pool_autoextend_percent.
	# Automatic extension requires dmeventd to be monitoring the LV.
	# 
	# Example
	# Using 70% autoextend threshold and 20% autoextend size, when a 1G
	# thin pool exceeds 700M, it is extended to 1.2G, and when it exceeds
	# 840M, it is extended to 1.44G:
	# thin_pool_autoextend_threshold = 70
	# 
	thin_pool_autoextend_threshold = 90

	# Configuration option activation/thin_pool_autoextend_percent.
	# Auto-extending a thin pool adds this percent extra space.
	# The amount of additional space added to a thin pool is this
	# percent of its current size.
	# 
	# Example
	# Using 70% autoextend threshold and 20% autoextend size, when a 1G
	# thin pool exceeds 700M, it is extended to 1.2G, and when it exceeds
	# 840M, it is extended to 1.44G:
	# thin_pool_autoextend_percent = 20
	# 
	thin_pool_autoextend_percent = 20
```

So, the LVM autoextend or the root pool should occur if you utilize  reach 18GB of usage in the default 20GB thinpool. You may not actually reach that before getting an error message that there's no room for a template,right? E.g. the kali template which is > 3GB.

When that autoextend activates, that would then add 4GB of space, giving you 24GB. The top 4GB is (presumably) inaccessible to the user until the next reboot when dom0 root's / is extended (via systemd script?), yes?

Brendan

**Comment by rustybird** [comment]
> oh, also once a template is downloaded, it has to be installed (essentially, unarchived)...does it need to write the contents to dom0 root filesystem _before_ migrating the content to the vm-pool?

Yep, that's when those large temporary files are written to /var/lib/qubes:

    $ rpm -ql qubes-template-fedora-34-4.0.6-202110020209
    [...]
    /var/lib/qubes/vm-templates/fedora-34/root.img
    /var/lib/qubes/vm-templates/fedora-34/root.img.part.00
    /var/lib/qubes/vm-templates/fedora-34/root.img.part.01
    /var/lib/qubes/vm-templates/fedora-34/root.img.part.02
    /var/lib/qubes/vm-templates/fedora-34/root.img.part.03
    /var/lib/qubes/vm-templates/fedora-34/root.img.part.04
    [...]

**Comment by Kannen** [comment]
I had a message saying that I had only 5% free space for varlibqubes.  So the 90% autoextend threshold was reached.

I also doubled the size of qubes-root and qubes-root--pool and it did not change the error message, and any trials to create a new qubes were failures. I have a lot of free space on my hard drive about 1TB.

**Comment by mockfox** [comment]
> I had a message saying that I had only 5% free space for varlibqubes. So the 90% autoextend threshold was reached.
> 
> I also doubled the size of qubes-root and qubes-root--pool and it did not change the error message, and any trials to create a new qubes were failures. I have a lot of free space on my hard drive about 1TB.

I managed to fix the issue on my install. I ran into the same problem with Qubes not recognizing the resize of qubes-root. I needed to run an additional update operation for the changes to be recognized.

```sudo lvresize --size 30g /dev/qubes_dom0/root
sudo resize2fs /dev/mapper/qubes_dom0-root```

**Comment by Kannen** [comment]
@mockfox Tx! it works, I had not increased the size of the filesystem after I extend the lv.

Maybe the lvm autoextend feature is not sufficient because the filesystem resize should be applied automatically too?

**Comment by brendanhoar** [comment]
> Maybe the lvm autoextend feature is not sufficient because the filesystem resize should be applied automatically too?

I concur - filesystem extension likely needs to be triggered if the pool size increased due to lvm autoextend. Likely needs a new systemd unit or similar?

B

**Comment by mockfox** [comment]
I think it would useful if there was a configuration option to set the size of `dom0-root`, eg in Qubes Global Settings.

**Comment by rustybird** [comment]
@kennethrrosen The varlibqubes pool is not really involved, it's just on the same filesystem as `/var/tmp` which is used during restore.
